File: TenisApi/Domain/Entities/LeagueMatch.cs
using System;
using System.Collections.Generic;

namespace TenisApi.Domain.Entities
{
    public class LeagueMatch
    {
        public int Id { get; private set; }
        public string Player1Name { get; private set; }
        public string Player2Name { get; private set; }
        public string MatchDate { get; private set; }
        public string? Score { get; private set; }
        public bool IsCompleted { get; private set; }
        public DateTime CreateDate { get; private set; }
        public bool IsDouble { get; private set; }
        public string? Player1PartnerName { get; private set; }
        public string? Player2PartnerName { get; private set; }
        public int? HostUserId { get; private set; }
        public int? GuestUserId { get; private set; }

        // Bire çok ilişki için veri havuzu (Domain Collection)
        private readonly List<MatchPointHistory> _pointHistories = new();
        public IReadOnlyCollection<MatchPointHistory> PointHistories => _pointHistories.AsReadOnly();

        // EF Core için gerekli boş kurucu metot
        #pragma warning disable CS8618 
        private LeagueMatch() { }
        #pragma warning restore CS8618

        public LeagueMatch(
            string player1Name, 
            string player2Name, 
            string matchDate, 
            bool isDouble = false, 
            string? player1PartnerName = null, 
            string? player2PartnerName = null,
            int? hostUserId = null,
            int? guestUserId = null)
        {
            if (string.IsNullOrWhiteSpace(player1Name))
                throw new ArgumentException("Oyuncu 1 adı boş olamaz.", nameof(player1Name));
            if (string.IsNullOrWhiteSpace(player2Name))
                throw new ArgumentException("Oyuncu 2 adı boş olamaz.", nameof(player2Name));
            if (string.IsNullOrWhiteSpace(matchDate))
                throw new ArgumentException("Maç tarihi boş olamaz.", nameof(matchDate));

            Player1Name = player1Name;
            Player2Name = player2Name;
            MatchDate = matchDate;
            IsCompleted = false;
            Score = null;
            CreateDate = DateTime.UtcNow;
            IsDouble = isDouble;
            Player1PartnerName = player1PartnerName;
            Player2PartnerName = player2PartnerName;
            HostUserId = hostUserId;
            GuestUserId = guestUserId;
        }

        // Maçı tamamlama iş kuralı (Domain Behavior)
        public void CompleteMatch(string score, IEnumerable<MatchPointHistory> pointHistories)
        {
            if (string.IsNullOrWhiteSpace(score))
                throw new ArgumentException("Tamamlanan bir maçın skoru girilmelidir.", nameof(score));

            Score = score;
            IsCompleted = true;

            _pointHistories.Clear();
            if (pointHistories != null)
            {
                _pointHistories.AddRange(pointHistories);
            }
        }

        // Maç tarihini güncelleme iş kuralı (Domain Behavior)
        public void RescheduleMatch(string newDate)
        {
            if (string.IsNullOrWhiteSpace(newDate))
                throw new ArgumentException("Yeni maç tarihi boş olamaz.", nameof(newDate));
            
            if (IsCompleted)
                throw new InvalidOperationException("Tamamlanmış bir maçın tarihi değiştirilemez.");

            MatchDate = newDate;
        }
    }
}

File: tenisdemo/Features/League/Models/LeagueMatch.swift
//
//  LeagueMatch.swift
//  tenisdemo
//
//  Created by Antigravity on 22.07.2026.
//

import Foundation

struct MatchPointHistory: Decodable, Identifiable {
    var id: Int { sequenceNumber }
    let p1Points: String
    let p2Points: String
    let p1Games: Int
    let p2Games: Int
    let p1Sets: Int
    let p2Sets: Int
    let server: String
    let sequenceNumber: Int
    let createdTime: String?
}

struct LeagueMatch: Identifiable, Decodable {
    let id: Int
    let player1Name: String
    let player2Name: String
    let matchDate: String
    let score: String?
    let isCompleted: Bool
    let createDate: String?
    let isDouble: Bool
    let player1PartnerName: String?
    let player2PartnerName: String?
    let pointHistories: [MatchPointHistory]?
    
    init(
        id: Int, 
        player1Name: String, 
        player2Name: String, 
        matchDate: String, 
        score: String?, 
        isCompleted: Bool, 
        createDate: String? = nil,
        isDouble: Bool = false,
        player1PartnerName: String? = nil,
        player2PartnerName: String? = nil,
        pointHistories: [MatchPointHistory]? = nil
    ) {
        self.id = id
        self.player1Name = player1Name
        self.player2Name = player2Name
        self.matchDate = matchDate
        self.score = score
        self.isCompleted = isCompleted
        self.createDate = createDate
        self.isDouble = isDouble
        self.player1PartnerName = player1PartnerName
        self.player2PartnerName = player2PartnerName
        self.pointHistories = pointHistories
    }
}
